pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,220 @@
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"sync"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+	dbs           map[string]*CouchDBStore
+	mux           sync.RWMutex
+}
+
+const (
+	blankHostErrMsg           = "hostURL for new CouchDB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+)
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client, dbs: map[string]*CouchDBStore{}}, nil
+}
+
+// CreateStore creates a new store with the given name.
+func (p *Provider) CreateStore(name string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	err := p.couchDBClient.CreateDB(context.Background(), name)
+
+	return err
+}
+
+// OpenStore opens an existing store with the given name and returns it.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	exists, err := p.storeExists(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, storage.ErrStoreNotFound
+	}
+
+	// Check cache first
+	store, exists := p.dbs[name]
+	if exists {
+		return store, nil
+	}
+
+	db := p.couchDBClient.DB(context.Background(), name)
+
+	if dbErr := db.Err(); dbErr != nil {
+		return nil, dbErr
+	}
+
+	store = &CouchDBStore{db: db}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+func (p *Provider) storeExists(name string) (bool, error) {
+	_, existsInCache := p.dbs[name]
+
+	if existsInCache {
+		return true, nil
+	}
+
+	return p.couchDBClient.DBExists(context.Background(), name)
+}
+
+// CloseStore closes a previously opened store.
+func (p *Provider) CloseStore(name string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return storage.ErrStoreNotFound
+	}
+
+	delete(p.dbs, name)
+
+	return store.db.Close(context.Background())
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close(context.Background())
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	p.dbs = make(map[string]*CouchDBStore)
+
+	return p.couchDBClient.Close(context.Background())
+}
+
+// CouchDBStore represents a CouchDB-backed database.
+type CouchDBStore struct {
+	db  *kivik.DB
+	mux sync.RWMutex
+}
+
+// Put stores the given key-value pair in the store.
+func (c *CouchDBStore) Put(k string, v []byte) error {
+	c.mux.Lock()
```

Comment:
You don't need a lock here since you're accessing an immutable field.

Reply:
Done. (I also got rid of that line in the close method that clears the cache by remaking the map, as discussed)